### PR TITLE
Implement ko/resolve

### PR DESCRIPTION
**Pull request by mattmoor**
The adds `resolve.ImageReferences` which takes serialized K8s yaml, a `build.Interface` and a `publish.Interface` and returns a serialized K8s yaml where references supported by the `build.Interface` have been built and published via the `publish.Interface`.

This also adds a `doc.go` for `ko`.

Progress towards: #80

### CODE REVIEW

File: ko/resolve/resolve.go [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+	for {
+		var obj interface{}
+		if err := decoder.Decode(&obj); err != nil {
+			if err == io.EOF {
+				return buf.Bytes(), nil
+			}
+			return nil, err
+		}
+
+		// Reflectively walk input, building and publishing supported references.
```
Comment (jonjohnsonjr) [inline ko/resolve/resolve.go]: Recursively?

File: ko/resolve/resolve.go [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+
+		if err := encoder.Encode(obj2); err != nil {
+			return nil, err
+		}
+	}
+}
+
+type replaceString func(string) (string, error)
+
+func replaceRecursive(obj interface{}, rs replaceString) (interface{}, error) {
```
Comment (jonjohnsonjr) [inline ko/resolve/resolve.go]: Can you add a comment describing what this does a little more. I'm struggling to parse this.

File: ko/resolve/resolve.go (hunk 1 above)
Comment (mattmoor) [reply ko/resolve/resolve.go]: Done.

File: ko/resolve/resolve.go (hunk 2 above)
Comment (mattmoor) [reply ko/resolve/resolve.go]: Added a very detailed comment.

File: ko/resolve/resolve.go [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+	"github.com/google/go-containerregistry/ko/build"
+	"github.com/google/go-containerregistry/ko/publish"
+)
+
+// ImageReferences resolves supported references to images within the input yaml
+// to published image digests.
+func ImageReferences(input []byte, builder build.Interface, publisher publish.Interface) ([]byte, error) {
+	// The loop is to support multi-document yaml files.
+	// This is handled by using a yaml.Decoder and reading objects until io.EOF, see:
+	// https://github.com/go-yaml/yaml/pull/301/files#diff-e51f89b39d976550831afb510a926be0R123
```
Comment (ImJasonH) [inline ko/resolve/resolve.go]: Can you link to the commit in the repo itself, instead of to a PR diff?

File: ko/resolve/resolve.go [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+		var obj interface{}
+		if err := decoder.Decode(&obj); err != nil {
+			if err == io.EOF {
+				return buf.Bytes(), nil
+			}
+			return nil, err
+		}
+
+		// Recursively walk input, building and publishing supported references.
+		// TODO(mattmoor): It may be worth considering gathering the supported references in
```
Comment (ImJasonH) [inline ko/resolve/resolve.go]: 👍 That seems much simpler.

File: ko/resolve/resolve.go [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+// it. This function does not support walking through struct types, but also
+// should not need to as the input is expected to be the result of parsing yaml
+// or json into an interface{}, which should only produce primitives, maps and
+// arrays. This function will return a copy of the object rebuilt by the walk
+// with the replacements made.
+func replaceRecursive(obj interface{}, rs replaceString) (interface{}, error) {
+	switch typed := obj.(type) {
+	case map[interface{}]interface{}:
+		sm := sync.Map{}
+		wg := sync.WaitGroup{}
```
Comment (ImJasonH) [inline ko/resolve/resolve.go]: supernit: I typically see this as `var wg sync.WaitGroup` (and `var sm sync.Map`)

File: ko/resolve/resolve_test.go [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+		bazRef: baz,
+	})
+	testHashes = map[string]v1.Hash{
+		fooRef: fooHash,
+		barRef: barHash,
+		bazRef: bazHash,
+	}
+)
+
+func noTestYAMLArrays(t *testing.T) {
```
Comment (ImJasonH) [inline ko/resolve/resolve_test.go]: `noTest` ?

File: ko/resolve/resolve.go (hunk 3 above)
Comment (mattmoor) [reply ko/resolve/resolve.go]: Done.

File: ko/resolve/resolve.go (hunk 4 above)
Comment (mattmoor) [reply ko/resolve/resolve.go]: Done.

File: ko/resolve/resolve.go (hunk 5 above)
Comment (mattmoor) [reply ko/resolve/resolve.go]: Done.

File: ko/resolve/resolve_test.go (hunk 6 above)
Comment (mattmoor) [reply ko/resolve/resolve_test.go]: Oops, done.

**Pull request merged**